### PR TITLE
Fire async error on panic in scorch async routine

### PR DESCRIPTION
**Pull request by pavelbazika**
Hi,

for yet unknown reason, my project suffer with corrupted indices on the storage. Slice dimensions are sometimes stored wrongly in zap files, but the error is not detected until read, so it is not easy to debug it. When such index is being read, it often results in go panic. I'm able to rebuild the corrupted index, so there's a workaround, but I need to catch such error first.

The error sometimes strikes during open, which can be handled by user (my) code easily. But sometimes it appears later, during asynchronous scorch mergerLoop operations for example. And here I'm currently helpless, panic in scorch internal goroutine makes my app crash.

Thus here is a pull request adding panic handling to persister, merger and introducer. I've found there's already an async error reporting mechanism via `fireAsyncError`, so I've used it.

I my code, I need to know which index failed, because I open multiple indices. Hence I created new error type which carries required information. I also considered passing whole scorch struct pointer, but I think that's too internal.

Another question is, what to do if there's no `onAsyncError` handler set. I prefer the panic to be silently catched and async task terminated (inspired by `parseMergePlannerOptions` handling). But the previous behaviour which leads to program crash can be preserved as well, if you think it is more appropriate.

Best regards

### CODE REVIEW

File: index/scorch/persister.go [hunk 1]
```diff
@@ -84,6 +84,15 @@ type notificationChan chan struct{}
 func (s *Scorch) persisterLoop() {
 	defer s.asyncTasks.Done()
```
Comment (abhinavdangeti) [inline index/scorch/persister.go:85]: This line can be moved into the defer func below.

Comment (pavelbazika) [reply index/scorch/persister.go]: Thanks, I moved it. I've also find a redundant call of Done below, but now I see it's already fixed in master.

File: index/scorch/introducer.go [hunk 2]
```diff
@@ -46,6 +46,16 @@ type epochWatcher struct {
 }
 
 func (s *Scorch) introducerLoop() {
+	defer func() {
+		if r := recover(); r != nil {
+			s.fireAsyncError(&AsyncPanicError{
+				Source: "introducer",
+				Path:   s.path,
+			})
+			s.asyncTasks.Done()
```
[the remark this answers is not in the document]
Comment (pavelbazika) [reply index/scorch/introducer.go]: Moved, but according to existing code, the call should be after fireAsyncError, thus after recover. See https://github.com/blevesearch/bleve/blob/2442216e141e62405f5939693a6837141b9b748d/index/scorch/merge.go#L36

File: index/scorch/merge.go [hunk 3]
```diff
@@ -29,6 +29,16 @@ import (
 )
 
 func (s *Scorch) mergerLoop() {
+	defer func() {
+		if r := recover(); r != nil {
+			s.fireAsyncError(&AsyncPanicError{
+				Source: "merger",
+				Path:   s.path,
+			})
+			s.asyncTasks.Done()
```
[the remark this answers is not in the document]
Comment (pavelbazika) [reply index/scorch/merge.go]: Ok, moved

**Review by abhinavdangeti — APPROVED** [review]

**Pull request merged**